### PR TITLE
Enable editing phone numbers and addresses

### PR DESCRIPTION
**Pull request by erikphansen**
## Description
This enables editing phone numbers and addresses in Profile 2.0. The bulk of the code came from existing legacy `Vet360` components used for Profile v1.

## Testing done
Local. Solid integration tests to follow so we can clean up and refactor some of this copy-pasta code.

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/applications/personalization/profile-2/components/personal-information/AddressesTable.jsx [hunk 1]
```diff
@@ -1,71 +1,26 @@
...
-
-import AddressView from 'platform/user/profile/vet360/components/AddressField/AddressView';
-import { FIELD_NAMES } from 'platform/user/profile/vet360/constants';
-import prefixUtilityClasses from 'platform/utilities/prefix-utility-classes';
+import MailingAddress from 'platform/user/profile/vet360/components/VAPMailingAddress';
+import ResidentialAddress from 'platform/user/profile/vet360/components/VAPResidentialAddress';
 
 import ProfileInfoTable from '../ProfileInfoTable';
 
-const ContactInfoCell = props => {
```
Comment (erikphansen) [inline src/applications/personalization/profile-2/components/personal-information/AddressesTable.jsx:12]: I was able to get rid of this fake read-only data I had put in place and replace each address field with a single `VAP Address` component

File: src/platform/user/profile/vet360/components/AddressField/AddressEditView.jsx [hunk 2]
```diff
@@ -0,0 +1,132 @@
+import React from 'react';
```
Comment (erikphansen) [inline src/platform/user/profile/vet360/components/AddressField/AddressEditView.jsx:1]: This component is similar to the `src/platform/user/profile/vet360/components/EmailField/EmailEditView.jsx` component. It is responsible for configuring and rendering a `VAPEditView` component. Most of this code was copied from the legacy `src/platform/user/profile/vet360/components/AddressField/AddressEditModal.jsx`

File: src/platform/user/profile/vet360/components/AddressField/VAPAddressField.jsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
+import React from 'react';
```
Comment (erikphansen) [inline src/platform/user/profile/vet360/components/AddressField/VAPAddressField.jsx:1]: This component is responsible for configuring and rendering a `VAPProfileField` to display and edit addresses

File: src/platform/user/profile/vet360/containers/AddressValidationView.jsx [hunk 4]
```diff
@@ -0,0 +1,334 @@
+import React from 'react';
```
Comment (erikphansen) [inline src/platform/user/profile/vet360/containers/AddressValidationView.jsx:1]: This file is very similar to the existing `src/platform/user/profile/vet360/containers/AddressValidationModal.jsx` but does not use a Modal component.

Comment (erikphansen) [reply src/platform/user/profile/vet360/containers/AddressValidationView.jsx:1]: I think we can do some refactoring in the entire address validation flow. For example, I'm pretty sure we can ignore any address suggestions we get back from the validation API that are not confirmed addresses.

Comment (erikphansen) [reply src/platform/user/profile/vet360/containers/AddressValidationView.jsx:1]: will definitely take a pass at cleaning this up when solid unit/integration tests are in place

File: src/platform/user/profile/vet360/tests/containers/AddressValidationView.unit.spec.jsx [hunk 5]
```diff
@@ -0,0 +1,417 @@
+import React from 'react';
```
Comment (erikphansen) [inline src/platform/user/profile/vet360/tests/containers/AddressValidationView.unit.spec.jsx:1]: mostly copied from the existing tests for the `AddressValidationModal` component.

File: src/platform/user/profile/vet360/containers/AddressValidationView.jsx [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+    state.vet360.addressValidation.addressValidationType;
+
+  return {
+    analyticsSectionName: VET360.ANALYTICS_FIELD_MAP[addressValidationType],
+    isLoading:
+      state.vet360.fieldTransactionMap[addressValidationType]?.isPending ||
+      isPendingTransaction(transaction),
+    addressValidationError:
+      state.vet360.addressValidation.addressValidationError,
+    suggestedAddresses: state.vet360.addressValidation.suggestedAddresses,
```
Comment (sanlouise) [inline src/platform/user/profile/vet360/containers/AddressValidationView.jsx:281]: Should we do optional chaining here?

File: src/platform/user/profile/vet360/containers/AddressValidationView.jsx [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+} from '../actions';
+import { getValidationMessageKey } from '../../utilities';
+import { ADDRESS_VALIDATION_MESSAGES } from '../../constants/addressValidationMessages';
+
+class AddressValidationView extends React.Component {
+  componentWillUnmount() {
+    focusElement(`#${this.props.addressValidationType}-edit-link`);
+  }
+
+  onChangeHandler = (address, selectedAddressId) => _event => {
```
Comment (sanlouise) [inline src/platform/user/profile/vet360/containers/AddressValidationView.jsx:34]: Why pass `event` here?

File: src/platform/user/profile/vet360/components/PhoneField/VAPPhoneField.jsx [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+      id,
+      countryCode,
+      extension,
+      phoneType,
+      inputPhoneNumber,
+      isTextable,
+      isTextPermitted,
+    } = value;
+
+    const strippedPhone = (inputPhoneNumber || '').replace(/[^\d]/g, '');
```
Comment (sanlouise) [inline src/platform/user/profile/vet360/components/PhoneField/VAPPhoneField.jsx:96]: Could we use a helper from `src/platform/forms/validations.js` here?

**Review by sanlouise — APPROVED** [review]

File: src/platform/user/profile/vet360/components/AddressField/AddressEditView.jsx [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+  /**
+   * Helper function that calls other helpers to:
+   * - totally remove data fields that are not set
+   * - set the form data's `view:livesOnMilitaryBase` prop to `true` if this is
+   *   an overseas military mailing address
+   *
+   * If the argument is not an object this function will simply return whatever
+   * was passed to it.
+   */
+  transformInitialFormValues = initialFormValues => {
```
[the remark this answers is not in the document]
Comment (sanlouise) [reply src/platform/user/profile/vet360/components/AddressField/AddressEditView.jsx:82]: It's actually probably fine as is, but it's hard to read - I like descriptive var names :) Just a preference

```
  transformInitialFormValues = initialFormValues => {
    if (!(initialFormValues instanceof Object)) {
      return initialFormValues;
    }
    let removedEmptyKeys = this.removeEmptyKeys(initialFormValues);
    return this.selectLivesOnMilitaryBaseCheckbox(removedEmptyKeys);
  };
```

**Review by bkjohnson — COMMENTED** [review]
I've left some comments already but some pieces I didn't take a thorough look at due to the size of this PR. Once you get a chance to look at these comments I will leave another review.

File: src/platform/user/profile/vet360/components/AddressField/AddressEditView.jsx [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+      </ContactInfoForm>
+    </div>
+  );
+
+  render() {
+    return (
+      <VAPEditView
+        getInitialFormValues={this.getInitialFormValues}
+        render={this.renderForm}
+        {...this.props}
```
[the remark this answers is not in the document]
Comment (erikphansen) [reply src/platform/user/profile/vet360/components/AddressField/AddressEditView.jsx:122]: I'm not a fan of doing prop spreading either. But this looks like a case where we actually do want to use it. ie, this functions as a HOC. We could try to be more selective about the prop passing one level higher, in the `VAPProfileField`, which is responsible for rendering this `AddressEditView` and passing props to it. But even at that level a lot of the work that the `VAPProfileField` does is to set up the functionality that allows for editing of profile data thus it needs to be passed down. At any rate, a lot of this refactoring is going to be done in [a follow up ticket](https://github.com/department-of-veterans-affairs/va.gov-team/issues/9564) after integration tests for Profile 2.0 are in place.

File: src/platform/user/profile/vet360/components/AddressField/VAPAddressField.jsx [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+import VAPProfileField from 'vet360/containers/VAPProfileField';
+
+import AddressEditView from './AddressEditView';
+import AddressValidationView from '../../containers/AddressValidationView';
+import AddressView from './AddressView';
+
+import { getFormSchema, getUiSchema } from './address-schemas';
+
+export const inferAddressType = (countryCodeIso3, stateCode) => {
```
[the remark this answers is not in the document]
Comment (erikphansen) [reply src/platform/user/profile/vet360/components/AddressField/VAPAddressField.jsx:22]: nope, good catch. there was a time with the previous AddressField component that this and `convertNextValueToCleanData` were used externally but that was changed months ago.

File: src/platform/user/profile/vet360/components/PhoneField/VAPPhoneField.jsx (hunk 8 above)
Comment (erikphansen) [reply src/platform/user/profile/vet360/components/PhoneField/VAPPhoneField.jsx:96]: the one in the shared validations is really for checking to see if a value is a valid phone number, while this is just to strip the non-numbers from a string. and I don't think it's worth making this a shared platform function just so we could reuse it here: https://github.com/department-of-veterans-affairs/vets-website/blob/fac0cd84ddd40bb45098ba269c7b64ffc50a6b2b/src/platform/forms/validations.js#L190

File: src/platform/user/profile/vet360/components/AddressField/AddressEditView.jsx [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+  selectLivesOnMilitaryBaseCheckbox = data => {
+    if (
+      data?.addressPou === ADDRESS_POU.CORRESPONDENCE &&
+      ADDRESS_DATA.militaryStates.includes(data?.stateCode) &&
+      ADDRESS_DATA.militaryCities.includes(data?.city)
+    ) {
+      return { ...data, 'view:livesOnMilitaryBase': true };
+    }
+    return data;
+  };
```
[the remark this answers is not in the document]
Comment (erikphansen) [reply src/platform/user/profile/vet360/components/AddressField/AddressEditView.jsx:71]: I'll meet you halfway on this; I still like having the logic live in its own function, but I like your idea of setting the `view:livesOnMilitaryBase` value inline if it's needed.

**Pull request merged**